key-export-client/src/lib.rs

```
@@ -0,0 +1,242 @@
+    pub fn new() -> Result<KeyExportContext, types::Error> {
+        let mut rng = rand_core::OsRng;
+        // Sample random 10 bytes to see that CSPRNG is available
+        let mut sample = [0u8; 10];
+        rng.try_fill_bytes(&mut sample)
+            .context("cryptographic randomness generator is not available")?;
+
+        Ok(KeyExportContext {
+            decryption_key: DecryptionKey::generate(&mut rng),
+        })
+    }
+
+    /// Returns a request body that needs to be sent to Dfns API in order to
+    /// export the key of the wallet with the given `wallet_id`.
+    ///
+    /// Throws `Error` in case of failure.
+    pub fn build_key_export_request(&self) -> Result<types::Request, types::Error> {
+        let req = KeyExportRequest {
+            supported_schemes: Vec::from(SUPPORTED_SCHEMES),
+            encryption_key: self.decryption_key.encryption_key(),
+        };
+        types::format_request(req)
+    }
+
+    /// Parses the response from Dfns API and recovers the private key.
+    ///
+    /// It returns the private key as a big endian byte array,
+    /// or an `Error` (if the private key cannot be recovered,
+    /// or is recovered but doesn’t match the public_key).
+    pub fn recover_secret_key(&self, response: String) -> Result<SecretKey, types::Error> {
+        // Parse response
+        let response: KeyExportResponse =
+            serde_json::from_str(&response).context("cannot parse key-export response")?;
+        // Parse and validate fields
+        let min_signers: u16 = response
+            .min_signers
+            .try_into()
+            .context("min_signers overflows u16")?;
+        let n: u16 = response
+            .encrypted_shares
+            .len()
+            .try_into()
+            .context("length of key_holders overflows u16")?;
+        if !(2 <= min_signers && min_signers <= n) {
+            return Err(new_error(
+                "threshold doesn't meet requirements: 2 <= min_signers <= n",
```

Requiring min_signers >= 2 seems a little ad hoc. Why not require min_signers >= 3? Or maybe in the future some customer who doesn't care about security (but does care about availability) does a 1-out-of-5 scheme?

---

I saw we had the same requirement in key-import, but probably this doesn't apply here.
Should I only keep "min_signers <= n" then @jkatzDfns ?

---

I think WASM module should accept any valid threshold, so it's $2 \le t \le n$. Any limitations that take into account the current deployment should be enforced by Coordinator and/or SDK

---

I assume that $t=1$ is not valid value for a threshold, because it defeats purpose of TSS by creating a SPOF at each signer. Also the math probably won't work well, for that reason $t=1$ is prohibited in cggmp21 and in the signers code.

---

I did not realize that t=1 is prohibited in signers

---

I guess it's fine to leave it, then. (FWIW, I still could imagine someday someone wants t=1 because they want to use our system but don't care about threshold.)

---

It means that signers share a polynomial of degree 0 (i.e. $F(x) = \text{sk}$). In the signing, threshold amount of signers (i.e. one) needs to be online. Protocol is not designed to support this case nor does it make any sense. Keygen - maybe, but signing should be different: each signer should do a regular signing when signature is requested (which is fast) instead of running CGGMP21 and trying to communicate to itself

---

I'd say we should push back that kind of requirement, because reducing SPOF is our security measure, and it's not compliant with decentralization vision of the future product (where no provider holds more than $t$ shares)

---

But it's only in the export functionality, I did not change the import.
Probably there will never exist a key with threshold 1 to export, for the reasons you mentioned, but is it a problem if the export code exports such keys as well?

---

I'm only saying that if we ever have $t=1$ keys, each "key share" would be just a key. It doesn't make sense to store "a key share" that contains a full key with a fields like other parties commitments (they are all the same) and shares indexes (they don't matter). Instead of key share, each signer would only store the full secret key. Export is trivial and does not require this WASM module, except maybe the encryption part.

---

I don't feel strongly about it, and we could always change it later if we ever have a customer who wants threshold 1. I just don't see why we are enforcing anything. The coordinator tells us what t and n are, so what case would there be where the coordinator tells us some values of t and n and we say "no, those values are not allowed"?

---

Not to mention the fact that currently we always have min_signers >= 3 ... so if we want error handling, then we should raise an error if that condition is ever violated

---

We should reject $t=1$ just because key export for those keys should be different

---

I agree that generally WASM should trust to coordinator on t,n values, but if WASM is called with $t=1$ then something wrong is going on

---

ok

---

@OrestisAlpos can you change back to discarding $t=1$?